File: Postgres/PostgresServer.m
// PostgresServer.m
//
// Created by Mattt Thompson (http://mattt.me/)
// Copyright (c) 2012 Heroku (http://heroku.com/)
// 
// Portions Copyright (c) 1996-2012, The PostgreSQL Global Development Group
// Portions Copyright (c) 1994, The Regents of the University of California
//
// Permission to use, copy, modify, and distribute this software and its
// documentation for any purpose, without fee, and without a written agreement
// is hereby granted, provided that the above copyright notice and this
// paragraph and the following two paragraphs appear in all copies.
//
// IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY FOR
// DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING
// LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
// EVEN IF THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
// SUCH DAMAGE.
//
// THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS ON AN
// "AS IS" BASIS, AND THE UNIVERSITY OF CALIFORNIA HAS NO OBLIGATIONS TO
// PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.


#import "PostgresServer.h"
#import "RecoveryAttempter.h"
#import <libpq-fe.h>


@interface PostgresServer()
@property BOOL isBusy;
@property (nonatomic) BOOL isRunning;
@end



@implementation PostgresServer

- (id)init {
	NSLog(@"%@: call -initWithName instead of -init", self.class.description);
	return nil;
}

- (id)initWithName:(NSString *)name version:(NSString *)version port:(NSUInteger)port varPath:(NSString *)varPath {
	self = [super init];
	if (self) {
		self.name = name;
		self.version = version;
		self.port = port;
		self.binPath = [BUNDLE_PATH stringByAppendingFormat:@"/Contents/Versions/%@/bin", self.version];
		self.varPath = varPath;
		self.runAtStartup = NO;
		self.stopAtQuit = YES;
		
		NSString *conf = [self.varPath stringByAppendingPathComponent:@"postgresql.conf"];
		if ([[NSFileManager defaultManager] fileExistsAtPath:conf]) {
			const char *t = [[NSString stringWithContentsOfFile:conf encoding:NSUTF8StringEncoding error:nil] UTF8String];
			for (int i = 0; t[i]; i++) {
				if (t[i] == '#')
					while (t[i] != '\n' && t[i]) i++;
				else if (strncmp(t+i, "port ", 5) == 0) {
					if (sscanf(t+i+5, "%*s %ld", &_port) == 1)
						break;
				}
			}
		}
		
	}
	
	return self;
}

- (id)initWithCoder:(NSCoder *)coder {
	self = [self initWithName:[coder decodeObjectForKey:@"name"]
					  version:[coder decodeObjectForKey:@"version"]
						 port:[coder decodeIntForKey:@"port"]
					  varPath:[coder decodeObjectForKey:@"varPath"]
			];
	
	if (self) {
		self.binPath = [coder decodeObjectForKey:@"binPath"];
		self.runAtStartup = [coder decodeBoolForKey:@"runAtStartup"];
		self.stopAtQuit = [coder decodeBoolForKey:@"stopAtQuit"];
	}
	return self;
}


- (void)encodeWithCoder:(NSCoder *)coder {
	[coder encodeObject:self.name forKey:@"name"];
	[coder encodeObject:self.version forKey:@"version"];
	[coder encodeInt:(unsigned)self.port forKey:@"port"];
	[coder encodeObject:self.binPath forKey:@"binPath"];
	[coder encodeObject:self.varPath forKey:@"varPath"];
	[coder encodeBool:self.runAtStartup forKey:@"runAtStartup"];
	[coder encodeBool:self.stopAtQuit forKey:@"stopAtQuit"];
}



#pragma mark - Async server control methods

- (void)startWithCompletionHandler:(PostgresServerControlCompletionHandler)completionBlock {
	
	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
		[self setIsBusyOnMainThread:YES];
		
		NSError *error = nil;
		PostgresDataDirectoryStatus dataDirStatus = [self statusOfDataDirectory:self.varPath error:&error];
		
		if (dataDirStatus == PostgresDataDirectoryEmpty) {
			BOOL serverDidInit = [self initDatabaseWithError:&error];
			if (!serverDidInit) {
				if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(NO, error); });
				return;
			}
			
			BOOL serverDidStart = [self startServerWithError:&error];
			if (!serverDidStart) {
				if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(NO, error); });
				return;
			}
			
			BOOL createdUser = [self createUserWithError:&error];
			if (!createdUser) {
				if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(NO, error); });
				return;
			}
			
			BOOL createdUserDatabase = [self createUserDatabaseWithError:&error];
			if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(createdUserDatabase, error); });
		}
		else if (dataDirStatus == PostgresDataDirectoryCompatible) {
			BOOL serverDidStart = [self startServerWithError:&error];
			if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(serverDidStart, error); });
		}
		else {
			if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(NO, error); });
		}
		
		[self setIsBusyOnMainThread:NO];
	});
}

- (void)stopWithCompletionHandler:(PostgresServerControlCompletionHandler)completionBlock {
	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
		[self setIsBusyOnMainThread:YES];
		
		NSError *error = nil;
		BOOL success = [self stopServerWithError:&error];
		if (completionBlock) dispatch_async(dispatch_get_main_queue(), ^{ completionBlock(success, error); });
		
		[self setIsBusyOnMainThread:NO];
	});
}




#pragma mark - Sync server control methods

- (PostgresDataDirectoryStatus)statusOfDataDirectory:(NSString *)dir error:(NSError **)outError {
	if (![[NSFileManager defaultManager] fileExistsAtPath:[dir stringByAppendingPathComponent:@"PG_VERSION"]]) {
		return PostgresDataDirectoryEmpty;
	}
	return PostgresDataDirectoryCompatible;
}

- (PostgresServerStatus)serverStatus {
	if (! [[NSFileManager defaultManager] fileExistsAtPath:self.binPath]) {
		return PostgresServerStatusNoBinDir;
	}
	
	NSTask *task = [[NSTask alloc] init];
	task.launchPath = [self.binPath stringByAppendingPathComponent:@"psql"];
	task.arguments = @[
						   [NSString stringWithFormat:@"-p%lu", self.port],
						   @"-A",
						   @"-q",
						   @"-t",
						   @"-c", @"SHOW data_directory;",
						   @"postgres"
	];
	
	NSPipe *outPipe = [[NSPipe alloc] init];
	task.standardOutput = outPipe;
	task.standardError = [[NSPipe alloc] init];
	
	[task launch];
	NSString *taskOutput = [[NSString alloc] initWithData:[[outPipe fileHandleForReading] readDataToEndOfFile] encoding:NSUTF8StringEncoding];
	[task waitUntilExit];
	
	NSString *expectedDataDirectory = self.varPath;
	NSString *actualDataDirectory = taskOutput.length>1 ? [taskOutput substringToIndex:taskOutput.length-1] : nil;
	
	switch(task.terminationStatus) {
		case 0:
			if (strcmp(actualDataDirectory.fileSystemRepresentation, expectedDataDirectory.fileSystemRepresentation) == 0) {
				[self setIsRunningOnMainThread:YES];
				return PostgresServerRunning;
			} else {
				[self setIsRunningOnMainThread:NO];
				return PostgresServerWrongDataDirectory;
			}
		case 2:
			[self setIsRunningOnMainThread:NO];
			return PostgresServerUnreachable;
		default:
			[self setIsRunningOnMainThread:NO];
			return PostgresServerStatusError;
	}
}

- (BOOL)startServerWithError:(NSError **)error {
	NSTask *task = [[NSTask alloc] init];
	task.launchPath = [self.binPath stringByAppendingPathComponent:@"pg_ctl"];
	task.arguments = @[
		/* control command          */ @"start",
		/* data directory           */ @"-D", self.varPath,
		/* wait for server to start */ @"-w",
		/* server log file          */ @"-l", self.logfilePath,
		/* allow overriding port    */ @"-o", [NSString stringWithFormat:@"-p %lu", self.port]
	];
	
	task.standardOutput = [[NSPipe alloc] init];
	task.standardError = [[NSPipe alloc] init];
	[task launch];
	NSString *controlTaskError = [[NSString alloc] initWithData:[[task.standardError fileHandleForReading] readDataToEndOfFile] encoding:NSUTF8StringEncoding];
	[task waitUntilExit];
	
	if (task.terminationStatus != 0 && error) {
		NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
		userInfo[NSLocalizedDescriptionKey] = NSLocalizedString(@"Could not start PostgreSQL server.",nil);
		userInfo[NSLocalizedRecoverySuggestionErrorKey] = controlTaskError;
		userInfo[NSLocalizedRecoveryOptionsErrorKey] = @[@"OK", @"Open Server Log"];
		userInfo[NSRecoveryAttempterErrorKey] = [[RecoveryAttempter alloc] init];
		userInfo[@"ServerLogRecoveryOptionIndex"] = @1;
		userInfo[@"ServerLogPath"] = self.logfilePath;
		*error = [NSError errorWithDomain:@"com.postgresapp.Postgres.pg_ctl" code:task.terminationStatus userInfo:userInfo];
	}

	if (task.terminationStatus == 0) {
		[self setIsRunningOnMainThread:YES];
	}
	
	return task.terminationStatus == 0;
}

- (BOOL)stopServerWithError:(NSError **)error {
	NSTask *task = [[NSTask alloc] init];
	task.launchPath = [self.binPath stringByAppendingPathComponent:@"pg_ctl"];
	task.arguments = @[
		/* control command         */ @"stop",
		/* fast mode               */ @"-m", @"f",
		/* data directory          */ @"-D", self.varPath,
		/* wait for server to stop */ @"-w",
	];
	
	task.standardOutput = [[NSPipe alloc] init];
	task.standardError = [[NSPipe alloc] init];
	[task launch];
	NSString *controlTaskError = [[NSString alloc] initWithData:[[task.standardError fileHandleForReading] readDataToEndOfFile] encoding:NSUTF8StringEncoding];
	[task waitUntilExit];
	
	if (task.terminationStatus != 0 && error) {
		NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
		userInfo[NSLocalizedDescriptionKey] = NSLocalizedString(@"Could not stop PostgreSQL server.",nil);
		userInfo[NSLocalizedRecoverySuggestionErrorKey] = controlTaskError;
		userInfo[NSLocalizedRecoveryOptionsErrorKey] = @[@"OK", @"Open Server Log"];
		userInfo[NSRecoveryAttempterErrorKey] = [[RecoveryAttempter alloc] init];
		userInfo[@"ServerLogRecoveryOptionIndex"] = @1;
		userInfo[@"ServerLogPath"] = self.logfilePath;
		*error = [NSError errorWithDomain:@"com.postgresapp.Postgres.pg_ctl" code:task.terminationStatus userInfo:userInfo];
	}
	
	if (task.terminationStatus == 0) {
		[self setIsRunningOnMainThread:NO];
	}
	
	return task.terminationStatus == 0;
}

- (BOOL)initDatabaseWithError:(NSError **)error {
	NSTask *task = [[NSTask alloc] init];
	task.launchPath = [self.binPath stringByAppendingPathComponent:@"initdb"];
	task.arguments = @[
		/* data directory */ @"-D", self.varPath,
		/* superuser name */ @"-U", @"postgres",
		/* encoding       */ @"--encoding=UTF-8",
		/* locale         */ @"--locale=en_US.UTF-8"
	];
	
	task.standardOutput = [[NSPipe alloc] init];
	task.standardError = [[NSPipe alloc] init];
	[task launch];
	NSString *initdbTaskError = [[NSString alloc] initWithData:[[task.standardError fileHandleForReading] readDataToEndOfFile] encoding:NSUTF8StringEncoding];
	[task waitUntilExit];
	
	if (task.terminationStatus != 0 && error) {
		NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
		userInfo[NSLocalizedDescriptionKey] = NSLocalizedString(@"Could not initialize database cluster.",nil);
		userInfo[NSLocalizedRecoverySuggestionErrorKey] = initdbTaskError;
		*error = [NSError errorWithDomain:@"com.postgresapp.Postgres.initdb" code:task.terminationStatus userInfo:userInfo];
	}
	
	return task.terminationStatus == 0;
}

- (BOOL)createUserWithError:(NSError **)error {
	NSTask *task = [[NSTask alloc] init];
	task.launchPath = [self.binPath stringByAppendingPathComponent:@"createuser"];
	task.arguments = @[
					   @"-U", @"postgres",
					   @"-p", @(self.port).stringValue,
					   @"--superuser",
					   NSUserName()
	];
	
	task.standardOutput = [[NSPipe alloc] init];
	task.standardError = [[NSPipe alloc] init];
	[task launch];
	NSString *taskError = [[NSString alloc] initWithData:[[task.standardError fileHandleForReading] readDataToEndOfFile] encoding:NSUTF8StringEncoding];
	[task waitUntilExit];
	
	if (task.terminationStatus != 0 && error) {
		NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
		userInfo[NSLocalizedDescriptionKey] = NSLocalizedString(@"Could not create default user.",nil);
		userInfo[NSLocalizedRecoverySuggestionErrorKey] = taskError;
		userInfo[NSLocalizedRecoveryOptionsErrorKey] = @[@"OK", @"Open Server Log"];
		userInfo[NSRecoveryAttempterErrorKey] = [[RecoveryAttempter alloc] init];
		userInfo[@"ServerLogRecoveryOptionIndex"] = @1;
		userInfo[@"ServerLogPath"] = self.logfilePath;
		*error = [NSError errorWithDomain:@"com.postgresapp.Postgres.createuser" code:task.terminationStatus userInfo:userInfo];
	}
	
	return task.terminationStatus == 0;
}

-(BOOL)createUserDatabaseWithError:(NSError **)error {
	NSTask *task = [[NSTask alloc] init];
	task.launchPath = [self.binPath stringByAppendingPathComponent:@"createdb"];
	task.arguments = @[
					   @"-p",
					   @(self.port).stringValue,
					   NSUserName()
	];
	
	task.standardOutput = [[NSPipe alloc] init];
	task.standardError = [[NSPipe alloc] init];
	[task launch];
	NSString *taskError = [[NSString alloc] initWithData:[[task.standardError fileHandleForReading] readDataToEndOfFile] encoding:NSUTF8StringEncoding];
	[task waitUntilExit];
	
	if (task.terminationStatus != 0 && error) {
		NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
		userInfo[NSLocalizedDescriptionKey] = NSLocalizedString(@"Could not create user database.",nil);
		userInfo[NSLocalizedRecoverySuggestionErrorKey] = taskError;
		userInfo[NSLocalizedRecoveryOptionsErrorKey] = @[@"OK", @"Open Server Log"];
		userInfo[NSRecoveryAttempterErrorKey] = [[RecoveryAttempter alloc] init];
		userInfo[@"ServerLogRecoveryOptionIndex"] = @1;
		userInfo[@"ServerLogPath"] = self.logfilePath;
		*error = [NSError errorWithDomain:@"com.postgresapp.Postgres.createdb" code:task.terminationStatus userInfo:userInfo];
	}
	
	return task.terminationStatus == 0;
}



#pragma mark - Async property helpers

- (void)setIsRunningOnMainThread:(BOOL)isRunning {
	dispatch_async(dispatch_get_main_queue(), ^{ self.isRunning = isRunning; });
}


- (void)setIsBusyOnMainThread:(BOOL)isBusy {
	dispatch_async(dispatch_get_main_queue(), ^{ self.isBusy = isBusy; });
}



#pragma mark - Custom properties

- (void)setIsRunning:(BOOL)isRunning {
	[self willChangeValueForKey:@"statusMessage"];
	_isRunning = isRunning;
	[self didChangeValueForKey:@"statusMessage"];
}


-(NSString *)logfilePath {
	return [self.varPath stringByAppendingPathComponent:@"postgres-server.log"];
}


- (NSString *)statusMessage {
	if (self.isRunning) {
		return [NSString stringWithFormat:@"PostgreSQL %@ – Running on port %lu", self.version, self.port];
	}
	else {
		return [NSString stringWithFormat:@"PostgreSQL %@ – Stopped", self.version];
	}
}


- (NSString *)statusMessageExtended {
	PostgresDataDirectoryStatus dataDirStatus = [self statusOfDataDirectory:self.varPath error:nil];
	switch (dataDirStatus) {
		case PostgresDataDirectoryIncompatible:
			return @"The selected data directory is not compatible with the selected server version.";
			break;
		case PostgresDataDirectoryCompatible:
			return @"Data directory compatible.";
			break;
		case PostgresDataDirectoryEmpty:
			return @"The data directory will be created the first time you start the server.";
			break;
		default:
			break;
	}
	
	return @"STATUS N/A";
}


- (NSArray *)databases {
	NSMutableArray *dbs = [[NSMutableArray alloc] init];
	
	NSString *connectionString = [NSString stringWithFormat:@"postgresql://:%lu", self.port];
	PGconn *conn = PQconnectdb(connectionString.UTF8String);
	PGresult *result = PQexec(conn, "SELECT datname FROM pg_database WHERE datallowconn ORDER BY LOWER(datname)");
	
	//ConnStatusType status = PQstatus(conn);
	
	for (int i=0; i<PQntuples(result); i++) {
		NSString *value = @(PQgetvalue(result, i, 0));
		[dbs addObject:value];
	}
	
	PQfinish(conn);
	
	return dbs;
}





@end

File: Postgres/PostgresServer.swift
//
//  PostgresServer.swift
//  Postgres
//
//  Created by Chris on 23/06/16.
//  Copyright © 2016 postgresapp. All rights reserved.
//

import Cocoa

class PostgresServer: NSObject, NSCoding {
	
	let BUNDLE_PATH = "/Applications/Postgres.app"
	
	enum DataDirectoryStatus {
		case Compatible
		case Incompatible
		case Empty
	}
	
	enum ServerStatus {
		case Startable
		case Running
		case NoBinDir
		case WrongDataDirectory
		case Error
	}
	
	enum ActionStatus {
		case Success
		case Failure(NSError)
	}
	
	
	dynamic var name: String = ""
	dynamic var version: String = ""
	dynamic var port: UInt = 0
	dynamic var binPath: String = ""
	dynamic var varPath: String = ""
	dynamic var runAtStartup: Bool = false
	dynamic var stopAtQuit: Bool = false
	
	dynamic private(set) var busy: Bool = false
	dynamic private(set) var statusMessage: String = ""
	dynamic private(set) var statusMessageExtended: String = ""
	dynamic private(set) var databases: [NSObject] = []
	
	dynamic private(set) var running: Bool {
		set {
			DispatchQueue.main.async { self._running = newValue }
		}
		get {
			return self._running
		}
	}
	private var _running: Bool = false {
		didSet {
			if self._running {
				self.statusMessage = "PostgreSQL \(self.version) - Running on port \(self.port)"
			} else {
				self.statusMessage = "PostgreSQL \(self.version) - Stopped"
			}
		}
	}
	
	dynamic var logfilePath: String {
		get {
			return self.varPath.appending("/postgres-server.log")
		}
	}
	
	var dataDirectoryStatus: DataDirectoryStatus {
		get {
			let pgVersionPath = self.varPath.appending("/PG_VERSION")
			if FileManager.default().fileExists(atPath: pgVersionPath) {
				do {
					let fileContents = try String(contentsOfFile: pgVersionPath)
					if fileContents.substring(to: fileContents.index(before: fileContents.endIndex)) == self.version {
						return .Compatible
					} else {
						return .Incompatible
					}
				} catch {}
			}
			return .Empty
		}
	}
	
	var serverStatus: ServerStatus {
		get {
			if !FileManager.default().fileExists(atPath: self.binPath) {
				return .NoBinDir
			}
			
			let task = Task()
			task.launchPath = self.binPath.appending("/psql")
			task.arguments = [
				"-p", String(self.port),
				"-A",
				"-q",
				"-t",
				"-c", "SHOW data_directory",
				"postgres"
			]
			let outPipe = Pipe()
			task.standardOutput = outPipe
			task.standardError = Pipe()
			task.launch()			
			let taskOutput = String(data: (outPipe.fileHandleForReading.readDataToEndOfFile()), encoding: .utf8) ?? "(incorrectly encoded error message)"
			task.waitUntilExit()
			
			switch task.terminationStatus {
			case 0:
				if taskOutput.characters.count > 0 && taskOutput.substring(to: taskOutput.index(before: taskOutput.endIndex)) == self.varPath {
					self.running = true
					return .Running
				}
				return .WrongDataDirectory
				
			case 2:
				self.running = false
				return .Startable
				
			default:
				self.running = false
				return .Error
			}
		}
	}
	
	
	convenience init(name: String, version: String, port: UInt, varPath: String) {
		self.init()
		
		self.name = name
		self.version = version
		self.port = port
		self.binPath = BUNDLE_PATH.appendingFormat("/Contents/Versions/%@/bin", self.version)
		self.varPath = varPath
		
		// TODO: read port from postgresql.conf
	}
	
	
	required convenience init(coder aDecoder: NSCoder) {
		self.init()
		
		guard let name = aDecoder.decodeObject(forKey: "name") as? String else { return }
		guard let version = aDecoder.decodeObject(forKey: "version") as? String else { return }
		guard let port = aDecoder.decodeObject(forKey: "port") as? UInt else { return }
		guard let varPath = aDecoder.decodeObject(forKey: "varPath") as? String else { return }
		
		self.name = name
		self.version = version
		self.port = port
		self.binPath = BUNDLE_PATH.appendingFormat("/Contents/Versions/%@/bin", self.version)
		self.varPath = varPath
		self.runAtStartup = aDecoder.decodeBool(forKey: "runAtStartup")
		self.stopAtQuit = aDecoder.decodeBool(forKey: "stopAtQuit")
	}
	
	
	func encode(with aCoder: NSCoder) {
		aCoder.encode(name, forKey: "name")
		aCoder.encode(version, forKey: "version")
		aCoder.encode(UInt(port), forKey: "port")
		aCoder.encode(binPath, forKey: "binPath")
		aCoder.encode(varPath, forKey: "varPath")
		aCoder.encode(runAtStartup, forKey: "runAtStartup")
		aCoder.encode(stopAtQuit, forKey: "stopAtQuit")
	}
	
	
	
	/*
	public async handlers
	*/
	func start(completionHandler: (_: ActionStatus) -> Void) {
		self.busy = true
		
		DispatchQueue.global().async {
			switch self.dataDirectoryStatus {
			
			case .Empty:
				let initRes = self.initDatabaseSync()
				if case .Failure = initRes {
					DispatchQueue.main.async {
						completionHandler(initRes)
					}
				}
				
				let startRes = self.startSync()
				if case .Failure = startRes {
					DispatchQueue.main.async {
						completionHandler(startRes)
					}
				}
				
				let createUserRes = self.createUserSync()
				if case .Failure = createUserRes {
					DispatchQueue.main.async {
						completionHandler(createUserRes)
					}
				}
				
				let createDBRes = self.createUserDatabaseSync()
				if case .Failure = createDBRes {
					DispatchQueue.main.async {
						completionHandler(createDBRes)
					}
				}
				
			case .Incompatible:
				let userInfo: [String: AnyObject] = [
					NSLocalizedDescriptionKey: NSLocalizedString("The data directory is not compatible with this version of PostgreSQL server.", comment: ""),
					NSLocalizedRecoverySuggestionErrorKey: "Please create a new Server."
				]
				let error = NSError(domain: "com.postgresapp.Postgres.data-directory", code: 0, userInfo: userInfo)
				DispatchQueue.main.async {
					completionHandler(.Failure(error))
				}
				
			case .Compatible:
				print(self.serverStatus)
				switch self.serverStatus {
				case .Running:
					let userInfo: [String: AnyObject] = [
						NSLocalizedDescriptionKey: "This PostgreSQL server is already running on port \(self.port)",
						NSLocalizedRecoverySuggestionErrorKey: "Please stop this server before starting again."
					]
					let error = NSError(domain: "com.postgresapp.Postgres.server-status", code: 0, userInfo: userInfo)
					DispatchQueue.main.async {
						completionHandler(.Failure(error))
					}
					
				case .NoBinDir:
					let userInfo: [String: AnyObject] = [
						NSLocalizedDescriptionKey: "The binaries for this PostgreSQL server were not found",
						NSLocalizedRecoverySuggestionErrorKey: "Create a new Server and try again."
					]
					let error = NSError(domain: "com.postgresapp.Postgres.server-status", code: 0, userInfo: userInfo)
					DispatchQueue.main.async {
						completionHandler(.Failure(error))
					}
					
				case .WrongDataDirectory:
					let userInfo: [String: AnyObject] = [
						NSLocalizedDescriptionKey: "There is already a PostgreSQL server running on port \(self.port)",
						NSLocalizedRecoverySuggestionErrorKey: "Please stop this server before.\n\nIf you want to use multiple servers, configure them to use different ports."
					]
					let error = NSError(domain: "com.postgresapp.Postgres.server-status", code: 0, userInfo: userInfo)
					DispatchQueue.main.async {
						completionHandler(.Failure(error))
					}
					
				case .Error:
					let userInfo: [String: AnyObject] = [
						NSLocalizedDescriptionKey: "Unknown error",
						NSLocalizedRecoverySuggestionErrorKey: ""
					]
					let error = NSError(domain: "com.postgresapp.Postgres.server-status", code: 0, userInfo: userInfo)
					DispatchQueue.main.async {
						completionHandler(.Failure(error))
					}
					
				case .Startable:
					let startRes = self.startSync()
					DispatchQueue.main.async {
						completionHandler(startRes)
					}
					
				}
			}
			
			DispatchQueue.main.async {
				self.busy = false
			}
		}
	}
	
	
	func stop(completionHandler: (_: ActionStatus) -> Void) {
		self.busy = true
		
		DispatchQueue.global().async {
			
			let stopRes = self.stopSync()
			DispatchQueue.main.async {
				completionHandler(stopRes)
			}
			
			DispatchQueue.main.async {
				self.busy = false
			}
		}
	}
	
	
	/*
	sync handlers
	*/
	private func startSync() -> ActionStatus {
		let task = Task()
		task.launchPath = self.binPath.appending("/pg_ctl")
		task.arguments = [
			"start",
			"-D", self.varPath,
			"-w",
			"-l", self.logfilePath,
			"-o", String("-p \(self.port)"),
		]
		task.standardOutput = Pipe()
		task.standardError = Pipe()
		task.launch()
		let errorDescription = String(data: (task.standardError?.fileHandleForReading.readDataToEndOfFile())!, encoding: .utf8) ?? "(incorrectly encoded error message)"
		task.waitUntilExit()
		
		if task.terminationStatus == 0 {
			self.running = true
			return .Success
		}
		else {
			let userInfo: [String: AnyObject] = [
				NSLocalizedDescriptionKey: NSLocalizedString("Could not start PostgreSQL server.", comment: ""),
				NSLocalizedRecoverySuggestionErrorKey: errorDescription,
				NSLocalizedRecoveryOptionsErrorKey: ["OK", "Open Server Log"],
				NSRecoveryAttempterErrorKey: ErrorRecoveryAttempter(recoveryAttempter: { (error, optionIndex) -> Bool in
					if optionIndex == 1 {
						NSWorkspace.shared().openFile(self.logfilePath, withApplication: "Console")
					}
					return true
				}),
			]
			let error = NSError(domain: "com.postgresapp.Postgres.pg_ctl", code: Int(task.terminationStatus), userInfo: userInfo)
			return .Failure(error)
		}
	}
	
	
	private func stopSync() -> ActionStatus {
		let task = Task()
		task.launchPath = self.binPath.appending("/pg_ctl")
		task.arguments = [
			"stop",
			"-m", "f",
			"-D", self.varPath,
			"-w",
		]
		task.standardOutput = Pipe()
		task.standardError = Pipe()
		task.launch()
		let errorDescription = String(data: (task.standardError?.fileHandleForReading.readDataToEndOfFile())!, encoding: .utf8) ?? "(incorrectly encoded error message)"
		task.waitUntilExit()
		
		if task.terminationStatus == 0 {
			self.running = false
			return .Success
		}
		else {
			let userInfo: [String: AnyObject] = [
				NSLocalizedDescriptionKey: NSLocalizedString("Could not stop PostgreSQL server.", comment: ""),
				NSLocalizedRecoverySuggestionErrorKey: errorDescription,
				NSLocalizedRecoveryOptionsErrorKey: ["OK", "Open Server Log"],
				NSRecoveryAttempterErrorKey: ErrorRecoveryAttempter(recoveryAttempter: { (error, optionIndex) -> Bool in
					if optionIndex == 1 {
						NSWorkspace.shared().openFile(self.logfilePath, withApplication: "Console")
					}
					return true
				})
			]
			let error = NSError(domain: "com.postgresapp.Postgres.pg_ctl", code: Int(task.terminationStatus), userInfo: userInfo)
			return .Failure(error)
		}
	}
	
	
	private func initDatabaseSync() -> ActionStatus {
		let task = Task()
		task.launchPath = self.binPath.appending("/initdb")
		task.arguments = [
			"-D", self.varPath,
			"-U", "postgres",
			"--encoding=UTF-8",
			"--locale=en_US.UTF-8"
		]
		task.standardOutput = Pipe()
		task.standardError = Pipe()
		task.launch()
		let errorDescription = String(data: (task.standardError?.fileHandleForReading.readDataToEndOfFile())!, encoding: .utf8) ?? "(incorrectly encoded error message)"
		task.waitUntilExit()
		
		if task.terminationStatus == 0 {
			return .Success
		}
		else {
			let userInfo: [String: AnyObject] = [
				NSLocalizedDescriptionKey: NSLocalizedString("Could not initialize database cluster.", comment: ""),
				NSLocalizedRecoverySuggestionErrorKey: errorDescription,
				NSLocalizedRecoveryOptionsErrorKey: ["OK", "Open Server Log"],
				NSRecoveryAttempterErrorKey: ErrorRecoveryAttempter(recoveryAttempter: { (error, optionIndex) -> Bool in
					if optionIndex == 1 {
						NSWorkspace.shared().openFile(self.logfilePath, withApplication: "Console")
					}
					return true
				})
			]
			let error = NSError(domain: "com.postgresapp.Postgres.initdb", code: Int(task.terminationStatus), userInfo: userInfo)
			return .Failure(error)
		}
	}
	
	
	private func createUserSync() -> ActionStatus {
		let task = Task()
		task.launchPath = self.binPath.appending("/createuser")
		task.arguments = [
			"-U", "postgres",
			"-p", String(self.port),
			"--superuser",
			NSUserName()
		]
		task.standardOutput = Pipe()
		task.standardError = Pipe()
		task.launch()
		let errorDescription = String(data: (task.standardError?.fileHandleForReading.readDataToEndOfFile())!, encoding: .utf8) ?? "(incorrectly encoded error message)"
		task.waitUntilExit()
		
		if task.terminationStatus == 0 {
			return .Success
		}
		else {
			let userInfo: [String: AnyObject] = [
				NSLocalizedDescriptionKey: NSLocalizedString("Could not create default user.", comment: ""),
				NSLocalizedRecoverySuggestionErrorKey: errorDescription,
				NSLocalizedRecoveryOptionsErrorKey: ["OK", "Open Server Log"],
				NSRecoveryAttempterErrorKey: ErrorRecoveryAttempter(recoveryAttempter: { (error, optionIndex) -> Bool in
					if optionIndex == 1 {
						NSWorkspace.shared().openFile(self.logfilePath, withApplication: "Console")
					}
					return true
				})
			]
			let error = NSError(domain: "com.postgresapp.Postgres.createuser", code: Int(task.terminationStatus), userInfo: userInfo)
			return .Failure(error)
		}
	}
	
	
	private func createUserDatabaseSync() -> ActionStatus {
		let task = Task()
		task.launchPath = self.binPath.appending("/createdb")
		task.arguments = [
			"-p", String(self.port),
			NSUserName()
		]
		task.standardOutput = Pipe()
		task.standardError = Pipe()
		task.launch()
		let errorDescription = String(data: (task.standardError?.fileHandleForReading.readDataToEndOfFile())!, encoding: .utf8) ?? "(incorrectly encoded error message)"
		task.waitUntilExit()
		
		if task.terminationStatus == 0 {
			return .Success
		}
		else {
			let userInfo: [String: AnyObject] = [
				NSLocalizedDescriptionKey: NSLocalizedString("Could not create user database.", comment: ""),
				NSLocalizedRecoverySuggestionErrorKey: errorDescription,
				NSLocalizedRecoveryOptionsErrorKey: ["OK", "Open Server Log"],
				NSRecoveryAttempterErrorKey: ErrorRecoveryAttempter(recoveryAttempter: { (error, optionIndex) -> Bool in
					if optionIndex == 1 {
						NSWorkspace.shared().openFile(self.logfilePath, withApplication: "Console")
					}
					return true
				})
			]
			let error = NSError(domain: "com.postgresapp.Postgres.createdb", code: Int(task.terminationStatus), userInfo: userInfo)
			return .Failure(error)
		}
	}
	
}




class DatabaseModel {
	
	var name: String = ""
	
}
